### PR TITLE
CB-5734 Datahub Autoscaling API Enhancements

### PR DESCRIPTION
**Pull request by smaniraju**
CB-5734 Enhancements
Added additional validations.
Added additional unit tests.
Removed AutoscalingMode and ScalingConfigurationRequest from DistroXAutoscaleClusterRequest.

### CODE REVIEW

File: autoscale/src/main/java/com/sequenceiq/periscope/service/AlertService.java [hunk 1]
```diff
@@ -132,12 +132,18 @@ public TimeAlert findTimeAlertByCluster(Long clusterId, Long alertId) {
...
         TimeAlert alert = timeAlertRepository.findByCluster(alertId, clusterId);
-        alert.setDescription(timeAlert.getDescription());
-        alert.setCron(timeAlert.getCron());
-        alert.setTimeZone(timeAlert.getTimeZone());
-        alert.setName(timeAlert.getName());
+        alert.setDescription(timeAlertForUpdate.getDescription());
+        alert.setCron(timeAlertForUpdate.getCron());
+        alert.setTimeZone(timeAlertForUpdate.getTimeZone());
+        alert.setName(timeAlertForUpdate.getName());
+        if (timeAlertForUpdate.getScalingPolicy() != null) {
```
Comment (sidseth) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/AlertService.java]: Is the ScalingPolicy field optional, or validated elsewhere?

File: autoscale/src/main/resources/schema/app/20200318083430_CB-5734_cluster_table_enhancements.sql [hunk 2]
```diff
@@ -5,8 +5,6 @@ ALTER TABLE cluster ADD COLUMN IF NOT EXISTS cb_stack_name VARCHAR(255);
 
 ALTER TABLE cluster ADD COLUMN IF NOT EXISTS cb_stack_type VARCHAR(255);
 
-ALTER TABLE cluster ADD COLUMN IF NOT EXISTS autoscale_mode VARCHAR(255);
```
Comment (sidseth) [inline autoscale/src/main/resources/schema/app/20200318083430_CB-5734_cluster_table_enhancements.sql]: This essentially means we are back to inefficient queries (full table scans), correct?

File: autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/v1/DistroXAutoScaleClusterV1Endpoint.java [hunk 3]
```diff
@@ -48,14 +51,14 @@
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = ClusterOpDescription.CLUSTER_SET_AUTOSCALE_STATE, produces = MediaType.APPLICATION_JSON, notes = DistroXClusterNotes.NOTES)
     DistroXAutoscaleClusterResponse updateAutoscaleConfigByClusterCrn(@PathParam("crn") String clusterCrn,
-            DistroXAutoscaleClusterRequest autoscaleClusterRequest);
+            @ValidDistroXAutoscaleRequest @Valid DistroXAutoscaleClusterRequest autoscaleClusterRequest) throws ParseException;
```
[the remark this answers is not in the document]
Comment (smaniraju) [reply autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/v1/DistroXAutoScaleClusterV1Endpoint.java]: As BAD_REQUEST

File: autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/validator/DistroXAutoscaleRequestValidator.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    duplicateAutoscaleHostGroups.toString());
+            com.sequenceiq.cloudbreak.validation.ValidatorUtil.addConstraintViolation(context, message, "hostGroup")
+                    .disableDefaultConstraintViolation();
+            return false;
+        }
+
+        List<AdjustmentType> invalidLoadAlertHostGroups =
+        request.getLoadAlertRequests().stream()
+                .map(loadAlertRequest -> loadAlertRequest.getScalingPolicy().getAdjustmentType())
+                .filter(adjustmentType -> !AdjustmentType.LOAD_BASED.equals(adjustmentType))
```
[the remark this answers is not in the document]
Comment (smaniraju) [reply autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/validator/DistroXAutoscaleRequestValidator.java]: Yes it needs to be specified to keep all alert definitions in sync.

File: autoscale/src/main/resources/schema/app/20200318083430_CB-5734_cluster_table_enhancements.sql (hunk 2 above)
Comment (smaniraju) [reply autoscale/src/main/resources/schema/app/20200318083430_CB-5734_cluster_table_enhancements.sql]: I have to investigate how to optimize this but I think we can avoid full table scans.

File: autoscale/src/main/java/com/sequenceiq/periscope/service/AlertService.java (hunk 1 above)
Comment (smaniraju) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/AlertService.java]: In the api it is not optional.

File: autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/validator/DistroXAutoscaleRequestValidator.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import javax.inject.Inject;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import com.sequenceiq.periscope.api.model.AdjustmentType;
+import com.sequenceiq.periscope.api.model.DistroXAutoscaleClusterRequest;
+import com.sequenceiq.periscope.api.util.ValidatorUtil;
+
+public class DistroXAutoscaleRequestValidator
```
[the remark this answers is not in the document]
Comment (smaniraju) [reply autoscale-api/src/main/java/com/sequenceiq/periscope/api/endpoint/validator/DistroXAutoscaleRequestValidator.java]: ack.

**Pull request merged**